### PR TITLE
feat: [Sparse Float Vector] segcore to support sparse vector search and get raw vector by id

### PR DESCRIPTION
**Pull request by zhengbuqian**
This PR adds the ability to search/get sparse float vectors in segcore, and added unit tests by modifying lots of existing tests into parameterized ones.

https://github.com/milvus-io/milvus/issues/29419

### CODE REVIEW

File: internal/core/src/query/PlanImpl.h [hunk 1]
```diff
@@ -64,19 +64,30 @@ struct Plan {
 struct Placeholder {
     std::string tag_;
     int64_t num_of_queries_;
-    int64_t line_sizeof_;
+    // TODO(SPARSE): add a dim_ field here, use the dim passed in search request
```
Comment (xiaofan-luan) [inline internal/core/src/query/PlanImpl.h:67]: is there a special reason we need this dim in search?

File: internal/core/src/segcore/FieldIndexing.cpp [hunk 2]
```diff
@@ -78,20 +83,78 @@ VectorFieldIndexing::GetDataFromIndex(const int64_t* seg_offsets,
...
+            auto dataset = knowhere::GenDataSet(rows, dim, mat);
+            dataset->SetIsSparse(true);
+            try {
+                if (chunk_id == 0) {
+                    index_->BuildWithDataset(dataset, conf);
+                } else {
+                    index_->AddWithDataset(dataset, conf);
+                }
+            } catch (SegcoreError& error) {
+                LOG_ERROR("growing sparse index build error: {}", error.what());
```
Comment (xiaofan-luan) [inline internal/core/src/segcore/FieldIndexing.cpp:130]: what happen if build or add error?

Will we rebuild everthing from chunk 0? will there be duplicated data?

File: internal/core/src/segcore/SegmentGrowingImpl.cpp [hunk 3]
```diff
@@ -603,6 +618,33 @@ SegmentGrowingImpl::bulk_subscript(FieldId field_id,
...
+SegmentGrowingImpl::bulk_subscript_sparse_float_vector_impl(
+    FieldId field_id,
+    const ConcurrentVector<SparseFloatVector>* vec_raw,
+    const int64_t* seg_offsets,
+    int64_t count,
+    milvus::proto::schema::SparseFloatArray* output) const {
+    AssertInfo(HasRawData(field_id.get()), "Growing segment loss raw data");
+
+    // if index has finished building index, grab from index
+    if (indexing_record_.SyncDataWithIndex(field_id)) {
```
Comment (xiaofan-luan) [inline internal/core/src/segcore/SegmentGrowingImpl.cpp:631]: could there be a race here?

1. we check syncdata and it is a not synced
2. insert thread build index and remove the chunk
3. when we try to retrieve from chunk, the original data has already been removed

Comment (xiaofan-luan) [reply internal/core/src/segcore/SegmentGrowingImpl.cpp:631]: I thought bulk_subscript_impl has same problem

File: internal/core/src/segcore/SegmentSealedImpl.cpp [hunk 4]
```diff
@@ -1514,14 +1544,17 @@ SegmentSealedImpl::generate_binlog_index(const FieldId field_id) {
     auto& field_index_meta = col_index_meta_->GetFieldIndexMeta(field_id);
     auto& index_params = field_index_meta.GetIndexParams();
 
+    bool is_sparse =
```
Comment (xiaofan-luan) [inline internal/core/src/segcore/SegmentSealedImpl.cpp:1547]: generate_binlog_index should rename to generate interim index.

File: internal/core/src/query/PlanImpl.h (hunk 1 above)
Comment (zhengbuqian) [reply internal/core/src/query/PlanImpl.h:67]: The main reason is to make sure the input for knowhere conforms to the API invariant: the `dataset->GetDim()` correctly returns the dim of the contained data, which is not yet true for those sparse query dataset. Though knowhere doesn't use the dim of the query dataset now.

**Comment by zhengbuqian** [comment]
UT for Cpp is failing due to https://github.com/milvus-io/milvus/issues/31154.

File: internal/core/src/segcore/FieldIndexing.cpp (hunk 2 above)
Comment (zhengbuqian) [reply internal/core/src/segcore/FieldIndexing.cpp:130]: this is related to both dense and sparse, I am updating it in https://github.com/milvus-io/milvus/pull/31170.

File: internal/core/src/segcore/SegmentGrowingImpl.cpp (hunk 3 above)
Comment (zhengbuqian) [reply internal/core/src/segcore/SegmentGrowingImpl.cpp:631]: this is related to both dense and sparse, I am updating it in https://github.com/milvus-io/milvus/pull/31170.

File: internal/core/src/segcore/SegmentSealedImpl.cpp (hunk 4 above)
Comment (zhengbuqian) [reply internal/core/src/segcore/SegmentSealedImpl.cpp:1547]: done

**Comment by zhengbuqian** [comment]
/unhold

#30357 has been merged and I have rebased this PR onto master, this PR now contains only 1 commit and is ready to merge. 

#35710 addresses 2 issues found during code review. those issues involves both dense/sparse thus using a separate PR to fix.

**Pull request merged**